packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -865,59 +877,92 @@ export class GarbageCollector implements IGarbageCollector {
     }
 
     /**
-     * Validates that all new references are correctly identified and processed. The basic principle for validation is
-     * that we should not have new references in the reference graph (GC data) that have not been notified to the
-     * garbage collector via `referenceAdded`.
-     * We validate that the references in the current reference graph should be a subset of the references in the last
-     * run's reference graph + references since the last run.
+     * Finds Missing Routes and logs a Performance event everytime a missing route is found
+     *
+     * Expected Routes = Routes in Current datastore graph and not in Previous datastore graph
+     * New Routes = New routes added by GC when a handle was added or root datastores
+     * Missing Routes = Expected Routes not in New Routes
+     *
+     * Node - represented as nodeId, it's a node on the GC graph
+     * Route - a reference with a parent, think `nodeId` & `reference`
+     * Reference - a node that is reachable from another node, think `nodeId`
+     * Outbound - a route or reference going from parent to child
+     * Graph - all nodes with their respective references
+     * Currently we are only filtering to get datastore references and routes
+     *
      * @param currentGCData - The GC data (reference graph) from the current GC run.
+     * @param previousGCData - The GC data (reference graph) from the previous GC run.
+     * @param newRoutes - New references added by GC from the previous to the current run.
+     * @returns - a list of missing routes
      */
-    private validateReferenceCorrectness(currentGCData: IGarbageCollectionData) {
+    private findMissingRoutes(
+        currentGCData: IGarbageCollectionData,
+        previousGCData: IGarbageCollectionData,
+        newRoutes: Map<string, string[]>,
+    ): [string, string[]][] {
         assert(
-            this.gcDataFromLastRun !== undefined,
+            previousGCData !== undefined,
             0x2b7, /* "Can't validate correctness without GC data from last run" */
         );
 
-        // Get a list of all the outbound routes (or references) in the current GC data.
-        const currentReferences: string[] = [];
-        for (const [nodeId, outboundRoutes] of Object.entries(currentGCData.gcNodes)) {
-            /**
-             * Remove routes from a child node to its parent which is added implicitly by the runtime. For instance,
-             * each adds its data store as an outbound route to mark it as referenced if the DDS is referenced.
-             * We won't get any explicit notification for these references so they must be removed before validation.
-             */
-            const explicitRoutes = outboundRoutes.filter((route) => !nodeId.startsWith(route));
-            currentReferences.push(...explicitRoutes);
-        }
+        const getOutboundReferences = (nodeId: string, references: string[]): string[] => {
+            if (references === undefined) {
```

Comment:
This check is not needed because the type of references is `string[]` so it won't be undefined.